lib/semantics/check-select-stmt.cpp

```diff
@@ -0,0 +1,316 @@
+          selectCaseIntValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(std::string strVal, parser::CharBlock source) {
+    auto selectCaseStringRangeValsIter = std::find_if(
+        selectCaseStringRangeVals_.begin(), selectCaseStringRangeVals_.end(),
+        [&strVal](const auto &mem) {
+          return (
+              (!mem.fromVal_ || (strVal.compare(mem.fromVal_.value()) >= 0)) &&
+              (!mem.toVal_ || (strVal.compare(mem.toVal_.value()) <= 0)));
+        });
+    if (selectCaseStringRangeValsIter != selectCaseStringRangeVals_.end()) {
+      ReportOverlapCaseValError(selectCaseStringRangeValsIter->source_, source);
+    }
+    auto selectCaseStringValsInsertStatus =
+        selectCaseStringVals_.insert(std::make_pair(strVal, source));
+    if (selectCaseStringValsInsertStatus.second == false) {
+      ReportOverlapCaseValError(
+          selectCaseStringValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(bool logicalVal, parser::CharBlock source) {
+    auto selectCaseLogicalValsInsertStatus =
+        selectCaseLogicalVals_.insert(std::make_pair(logicalVal, source));
+    if (selectCaseLogicalValsInsertStatus.second == false) {
+      ReportOverlapCaseValError(
+          selectCaseLogicalValsInsertStatus.first->second, source);
+    }
+  }
+  void Insert(std::optional<std::int64_t> lowerVal,
+      std::optional<std::int64_t> upperVal, parser::CharBlock source) {
+    SelectCaseRangeType<std::int64_t> intRangeValue(lowerVal, upperVal, source);
+    auto selectCaseIntValsIter = std::find_if(selectCaseIntVals_.begin(),
+        selectCaseIntVals_.end(), [&intRangeValue](const auto &mem) {
+          return ((!intRangeValue.fromVal_ ||
+                      (intRangeValue.fromVal_.value() <= mem.first)) &&
+              (!intRangeValue.toVal_ ||
+                  (intRangeValue.toVal_.value() >= mem.first)));
+        });
+    if (selectCaseIntValsIter != selectCaseIntVals_.end()) {
+      ReportOverlapCaseValError(selectCaseIntValsIter->second, source);
+    }
+    auto selectCaseIntRangeValsIter = std::find_if(
+        selectCaseIntRangeVals_.begin(), selectCaseIntRangeVals_.end(),
+        [&intRangeValue](const auto &mem) {
+          return ((!mem.fromVal_ && !intRangeValue.fromVal_) ||
+              (!mem.toVal_ && !intRangeValue.toVal_) ||
+              ((!mem.fromVal_ ||
+                   (intRangeValue.fromVal_ &&
+                       (intRangeValue.fromVal_.value() >= mem.fromVal_))) &&
+                  (!mem.toVal_ ||
+                      (intRangeValue.fromVal_ &&
+                          (intRangeValue.fromVal_.value() <= mem.toVal_)))) ||
+              ((!mem.fromVal_ ||
+                   (intRangeValue.toVal_ &&
+                       (intRangeValue.toVal_.value() >= mem.fromVal_))) &&
+                  (!mem.toVal_ ||
+                      (intRangeValue.toVal_ &&
+                          (intRangeValue.toVal_.value() <= mem.toVal_)))));
```

Comment:
Perhaps some of these complicated predicates could be simplified by creating functions that check if a value is in the range of a SelectCaseRangeType.

Reply:
I have abstracted all the common code and created template functions for the same.